### ISSUE TITLE
Triggers behave unexpectedly (was: "when using locals") because two trigger instances from the same source line can't coexist

### ISSUE THREAD

**Issue by cdglasz** [KSP-KOS/KOS#2272]
I'm trying to create triggers for each science module with a loop, and running into some trouble. The following should print the numbers 1 through 10 (not necessarily in order, I'm not sure if triggers are always checked in order of creation?), but it doesn't. Now, if the use of locals was totally broken in triggers, you might expect that it would print "10" ten times, or maybe do nothing at all. 
Instead, none of those things happen. It just prints "1", once.
```
{
  local trig to false.
  local a to 0.
  until a > 10 {
    local b to a.
    on trig {
      print b.
    }
  }
  set trig to true.
}
```
Here's another case, this time with a `for` loop.
```
set offsets to list(1,2,3,4,5).
for t in offsets {
	local offset to t.
  local utime to time:seconds + offset.
  when time:seconds > utime then {
    print "t: " + t + "  offset: " + offset.
  }
}
wait until 0.
```
This should print something every second for five seconds. At worst, you might expect it to wait five seconds, then print five lines. Instead, it simply prints `t: 5  offset: 1`.

**Comment by cdglasz** [comment]
After some further digging, this has nothing to do with loops.
```
set trig to false.

// A function to print something each time the trigger condition is checked
function verbose_condition {
	parameter b.
	return {
		print "Checking condition for " + b + ": " + trig.
		return trig.
	}.
}

// First, try creating triggers the normal way. This works.
{
	local b to "outside 1".
	local cond to verbose_condition(b).
	on cond() {
		print "Trigger condition met: " + b.
	}
}
{
	local b to "outside 2".
	local cond to verbose_condition(b).
	on cond() {
		print "Trigger condition met: " + b.
	}
}

// Now do it with a function. Only the first one continues to be checked.
function create_trigger {
	parameter b.
	local cond to verbose_condition(b).
	on cond() {
		print "Trigger condition met: " + b.
	}
}

create_trigger("function 1").
create_trigger("function 2").

wait 0.1.
set trig to true.
wait 0.1.
```
This reveals that only the first trigger created in a function continues to be checked. The second one is checked exactly once, when it's created, and then is never printed again.

**Comment by Dunbaratu** [comment]
I don't think it has anything to do with loops or with closures or any of that.

I think what is actually happening is that you're assuming that running the code containing ``on cond()`` twice causes two different triggers to get created, when it doesn't.  The uniqueness of a trigger is decided at compile time based on where the code is compiled, like so, not based on how many times the definition got executed:

The "Equals" check for a trigger is this, from ``kOS.Safe.Execution.TriggerInfo``:

```
    /// <summary>
    /// Two TriggerInfos shall be considered equivalent (and thus disallowed
    /// from existing simultaneously in collections that demand uniqueness) if
    /// they are non-callback triggers that refer to the same entry
    /// point.  For callback triggers, ANY two callbacks that are not the
    /// same reference shall be consided unequal.
    /// The closure field is not compared because it should be hypothetically
    /// impossible for two triggers that call the same IP location to have
    /// differing closures unless we really did something wrong when we created
    /// those closures in the first place.
    /// </summary>
    /// <param name="other"></param>
    /// <returns></returns>
    public override bool Equals(object other)
    {
        TriggerInfo otherInfo = other as TriggerInfo;
        if (otherInfo == null)
            return false;

        if (IsCSharpCallback)
        {
            // For callbacks, only reference-equals are good enough.
            // Hypothetically the same chunk of user code (same entry point)
            // could have been given as a callback to two different areas of
            // our C# code, and if so we still want them to be two different
            // callbacks tracked separately.
            if (otherInfo != this)
                return false;
        }
        else
        {
            // for non-callbacks, as long as they refer to the same
            // chunk of user code, they are equal:
            if (EntryPoint != otherInfo.EntryPoint)
                return false;
        }

        return true;
    }
```

The trigger that ``create_trigger("function 2")`` tries to add is considered *equal* to the trigger that ``create_trigger("function 1")`` already added.  So it says, "That trigger has already been queued up so I shouldn't insert it into the queue again."

I'm not sure I'd call this a bug, more like "not a use case it was designed to support".  It may be possible to change it, but it wasn't in the original plans to be queuing up an additional instance of the same trigger while a previous instance is still pending.  Allowing it will create a danger of it being possible to queue up pending trigger calls faster than they get executed, making the system spend forever stuck in interrupt mode, never getting back to mainline code.  (Preventing this is why that check is there.)

This is the scenario that that 'forced uniquenss' in the queued triggers is designed to prevent:

1 - At the top of each FixedUpdate tick (timeslice, if you will), the kOS system always interrupts what it was doing and inserts calls to all the enabled ``on`` and ``when`` triggers.  These triggers either decide that the condition isn't right, in which case they abort early and return, or they decide the condition was right, so they continue on into executing the body and return at the bottom of the body.
2 - When all these interrupt calls return, kOS should be right back where it was, with the stack in the same state it was at the end of the previous tick, so now it continues to executes mainline code again right where it left off.
3 - But now imagine a user makes a trigger which has a very big conditional check at the top that requires a lot of execution work to get the answer.  (Imagine if ``cond()`` in your ``on cond()`` was a function call that contained a loop that took a lot of of IPU's to finish, like a walk of the entire ship parts tree).
4 - Because kOS inserts a new call to the trigger condition check at the start of every update, and it's taking longer than one update for the previous call to finish, the mainline code will *never* happen.  kOS is inserting calls to perform ``if cond() <> cond() from the prev iteration`` faster than they can be executed.  The system never returns from the interruptions it's inserting.

The way this was prevented was to have logic that demands that once a trigger is already pending to be checked, (has already had a call inserted onto the call stack), then it cannot get a second instance of itself inserted again until the first instance has been executed.  Thus at least *some* mainline code ends up executing before the condition gets checked again.

The problem here, is that it sees your two calls to execute ``on cond()`` as being attempts to insert the same trigger faster than it's getting executed, because it's not seeing them as being two different triggers.

**Comment by Dunbaratu** [comment]
It may be possible to change this behaviour by distinguishing between the triggers that insert because your code executed ``on`` or ``when`` again versus the triggers that get inserted because kOS automatically inserted them at the start of a tick.  i.e. treat a re-execution of the same ``on`` or ``when`` command as a new unique trigger by adding an additional field to that ``Equals()`` check.

**Comment by Dunbaratu** [comment]
I'm going to create a new issue to more accurately describe the problem under the hood, and link it to this one.

**Comment by Dunbaratu** [comment]
I edited the title to more clearly describe the actual problem.  It will make associating this with a PR easier.

**Comment by Dunbaratu** [comment]
It's fuzzy whether to call it fixing a "bug" or implementing a new "enhancement", but I decided to call it a "bug" because the original intent wasn't well advertised to the user so it was reasonable to assume it would work in the desired new way.

**Comment by Dunbaratu** [comment]
Progress report - I made it capable of having more than one active trigger at a time. 
But this revealed a problem with ON, being that when comparing "is current value different than the prev value", it's storing only one prev value for all instances of the same ON in the code.

And looking at that indicated to me that the only way to fix it is going to be to make ON and WHEN have local scope inside their boolean checks at the top (right now they don't) so I can make complex logic using local variables instead of doing everything on the stack I figured out that it's literally impossible to DUP and SWAP in the right way to get the things in the position I need to do this logic at the top:

Get prev value, get new value, store new value where old value was, compare to prev value, branch based on result.,

That used to be easy with a pre-determined at compile time identifier for it all.  Now that I have to wait till runtime to store things, I really need that local scope, and that's going to take more work.
